azure-iot-device/tests/common/pipeline/pipeline_ops_test.py

```diff
@@ -0,0 +1,461 @@
+            assert handle_exceptions.handle_background_exception.call_count == 1
+            assert (
+                type(handle_exceptions.handle_background_exception.call_args[0][0])
+                is pipeline_exceptions.OperationError
+            )
+
+        @pytest.mark.it(
+            "Triggers and removes callbacks that have been added to the operation according to LIFO order"
+        )
+        def test_trigger_callbacks(self, mocker, cls_type, init_kwargs, error):
+            mocker.spy(handle_exceptions, "handle_background_exception")
+
+            # Set up callback mocks
+            cb1_mock = mocker.MagicMock()
+            cb2_mock = mocker.MagicMock()
+            cb3_mock = mocker.MagicMock()
+
+            def cb1(op, error):
+                # Callback 2 and 3 have already been triggered
+                assert cb1_mock.call_count == 1
+                assert cb2_mock.call_count == 1
+                assert cb3_mock.call_count == 1
+                assert len(op.callbacks) == 0
+
+            def cb2(op, error):
+                # Callback 3 has already been triggered, but Callback 1 has not
+                assert cb1_mock.call_count == 0
+                assert cb2_mock.call_count == 1
+                assert cb3_mock.call_count == 1
+                assert len(op.callbacks) == 1
+
+            def cb3(op, error):
+                # Callback 3 has been triggered, but no others have been.
+                assert cb1_mock.call_count == 0
+                assert cb2_mock.call_count == 0
+                assert cb3_mock.call_count == 1
+                assert len(op.callbacks) == 2
+
+            cb1_mock.side_effect = cb1
+            cb2_mock.side_effect = cb2
+            cb3_mock.side_effect = cb3
+
+            # Attach callbacks to op
+            init_kwargs["callback"] = cb1_mock
+            op = cls_type(**init_kwargs)
+            op.add_callback(cb2_mock)
+            op.add_callback(cb3_mock)
+            assert len(op.callbacks) == 3
+            assert not op.completed
+
+            # Run the completion
+            op.complete(error=error)
+
+            assert op.completed
+            assert cb3_mock.call_count == 1
+            assert cb3_mock.call_args == mocker.call(op=op, error=error)
+            assert cb2_mock.call_count == 1
+            assert cb2_mock.call_args == mocker.call(op=op, error=error)
+            assert cb1_mock.call_count == 1
+            assert cb1_mock.call_args == mocker.call(op=op, error=error)
+
+            # Because exceptions raised in callbacks are caught and sent to the background exception handler,
+            # the assertions in the above callbacks won't be able to directly raise AssertionErrors that will
+            # allow for testing normally. Instead we should check the background_exception_handler to see if
+            # any of the assertions raised errors and sent them there.
+            assert handle_exceptions.handle_background_exception.call_count == 0
```

Comment:
>assert handle_exceptions.handle_background_exception.call_count == 0 [](start = 12, length = 68)

This assumption may be correct but it's an assumption nonetheless.  In other words, you assume the asserts in cb1, cb2, and cb3 all pass because there are no calls to handle_background_exception, but you don't verify this.  If I comment out `cb1_mock.side_effect = cb1`, this test would still pass without ever running the checks inside cb1.  Likewise if I break the exception handling around the callback invocation, it's possible that those asserts could fail, but the test would pass.

Reply:
Perhaps the comment is unclear - doing an assertion raises an `AssertionError`. The `AssertionError` normally is able to bubble up because no Exception handling accounts for it, until it is intercepted by the testing framework, which uses that to fail the test. However, due to us catching ALL exceptions that come from a callback, the AssertionError is also sent to the background handler - like any other exception would be.

I agree that if you comment out the mock side effect lines, this test will be broken, but that's because that's a crucial line of the test. You could say that about MANY lines of tests. So long as the mock side effects ARE set up, this test works correctly.

To your second point, if the Exception handling around the callback invocation were broken within the `.complete()` method, then this test would still fail, as suddenly the `AssertionError`s raised by the callback would be able to propagate to the test framework and immediately cause a failure.

---
In reply to: [346502193](https://github.com/Azure/azure-iot-sdk-python/pull/392#discussion_r346502193) [](ancestors = 346502193)